staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -161,33 +179,43 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		defer utilruntime.HandleCrash()
 		<-stopCh
 
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
 		if client != nil {
 			client.Close()
 			clientErr = fmt.Errorf("server is shutting down")
 		}
 	}()
 
+	// limit to a request every half of the configured timeout with a maximum burst of one
+	limiter := rate.NewLimiter(rate.Every(timeout/2), 1)
+	// rate limited requests will be served from the cache
+	lastError := &atomicError{}
+
 	return func() error {
 		// Given that client is closed on shutdown we hold the lock for
 		// the entire period of healthcheck call to ensure that client will
 		// not be closed during healthcheck.
 		// Given that healthchecks has a 2s timeout, worst case of blocking
 		// shutdown for additional 2s seems acceptable.
-		lock.Lock()
-		defer lock.Unlock()
+		lock.RLock()
+		defer lock.RUnlock()
 		if clientErr != nil {
+			lastError.Store(clientErr)
 			return clientErr
 		}
+		if limiter.Allow() == false {
+			return lastError.Load()
+		}
 		ctx, cancel := context.WithTimeout(context.Background(), timeout)
 		defer cancel()
 		// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
 		_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
-		if err == nil {
-			return nil
+		if err != nil {
+			err = fmt.Errorf("error getting data from etcd: %w", err)
 		}
-		return fmt.Errorf("error getting data from etcd: %w", err)
+		lastError.Store(err)
```

Comment:
I'm trying to decide how big of a deal it is that a slow failed response to etcd can overwrite a fast successful response from etcd that started later:

1. client X hits /healthz, triggers request A to etcd
2. client Y hits /healthz, triggers request B to etcd
3. request B completes successfully, stores nil
4. request A completes unsuccessfully (maybe was to a dead etcd member, took 2 seconds, timed out, etc), stored error
5. client Z hits /healthz, hits rate limiter, gets cached error response B

Should we remember the time at which we started the `client.Get` request, pass it to the Store function, and make storing conditional on not overwriting a result that started later?

Reply:
I've added a test case, the use case is legit, but I think is going to be very rare because the rateLimiter depends on the healthzcheck timeout, the window is very small ... but definitively possible and is easy to cover it

Reply:
check the test  I've added, is easier to see there what  I mean